### PR TITLE
Add thrust support for nms

### PR DESCRIPTION
**Pull request by Laurawly**
@kazum @masahi @icemelon9 Please feel free to raise any suggestions or comments.

### CODE REVIEW

File: src/runtime/contrib/thrust/thrust.cu [hunk 1]
```diff
@@ -65,6 +65,105 @@ void thrust_sort(DLTensor* input,
   }
 }
 
+// Performs sorting along axis -1 and returns both sorted values and indices.
+template<typename DataType, typename IndicesType>
+void thrust_sort_nms(DLTensor* input,
+                 DLTensor* valid_count,
```
Comment (masahi) [inline src/runtime/contrib/thrust/thrust.cu]: Align the indent

**Comment by masahi** [comment]
@Laurawly Can we refactor `thrust_sort` and `thrust_sort_nms`? I see considerable duplicate.

File: src/runtime/contrib/thrust/thrust.cu [hunk 2]
```diff
@@ -65,6 +65,105 @@ void thrust_sort(DLTensor* input,
...
+  int n_values = input->shape[input->ndim - 1];
+  int n_iter = 1;
+  for (int i = 0; i < input->ndim - 1; ++i) {
+    n_iter *= input->shape[i];
+  }
+
+  thrust::copy(data_ptr, data_ptr + n_iter * n_values, values_ptr);
+
+  for (int i = 0 ; i < n_iter; ++i) {
+    int current_values = valid_count_ptr[i];
```
Comment (masahi) [inline src/runtime/contrib/thrust/thrust.cu]: if we replace `current_values` with `n_values`, the sorting code above and below looks identical with `thrust_sort`

File: src/runtime/contrib/thrust/thrust.cu [hunk 3]
```diff
@@ -65,6 +65,105 @@ void thrust_sort(DLTensor* input,
...
+    } else if (out_dtype == "float64") {
+      thrust_sort_nms<int64_t, double>(input, valid_count, values_out, indices_out, is_ascend);
+    } else {
+      LOG(FATAL) << "Unsupported output dtype: " << out_dtype;
+    }
+  } else {
+    LOG(FATAL) << "Unsupported input dtype: " << data_dtype;
+  }
+});
+
```
Comment (masahi) [inline src/runtime/contrib/thrust/thrust.cu]: Likely we can share this if/else with `thrust_sort` using template or lambda

File: topi/python/topi/cuda/sort.py [hunk 4]
```diff
@@ -237,6 +237,51 @@ def sort_nms_ir(data, valid_count, output, axis, is_ascend):
...
+    valid_count_buf = tvm.tir.decl_buffer(valid_count.shape, valid_count.dtype,
+                                          "valid_count_buf", data_alignment=4)
+    out_bufs = [
+        tvm.tir.decl_buffer(data.shape, data.dtype, "value_buf", data_alignment=8),
+        tvm.tir.decl_buffer(data.shape, "int32", "indices_buf", data_alignment=8)
+    ]
+    out = te.extern([data.shape, data.shape],
+                    [data, valid_count],
+                    lambda ins, outs: tvm.tir.call_packed(
+                        "tvm.contrib.thrust.sort_nms", ins[0], ins[1], outs[0], outs[1], is_ascend),
```
Comment (kazum) [inline topi/python/topi/cuda/sort.py]: thrust_sort_nms() supports only axis -1.
Should we add `assert(axis == -1)`?

File: src/runtime/contrib/thrust/thrust.cu (hunk 2 above)
Comment (Laurawly) [reply src/runtime/contrib/thrust/thrust.cu]: `thrust_sort_nms` needs one more input tensor than `thrust_sort`, I'm not sure how to pass the `valid_count` tensor without creating a new function.

File: src/runtime/contrib/thrust/thrust.cu (hunk 3 above)
Comment (Laurawly) [reply src/runtime/contrib/thrust/thrust.cu]: Could you elaborate more on this?

File: src/runtime/contrib/thrust/thrust.cu (hunk 2 above)
Comment (masahi) [reply src/runtime/contrib/thrust/thrust.cu]: I imagine we can create another function `thrust_sort_common`, and put all common code in it. Then have existing `thrust_sort` and `thrust_sort_nms` dispatch into it appropriately.

Comment (masahi) [reply src/runtime/contrib/thrust/thrust.cu]: `thrust_sort_common` should take a function (lambda) as an argument, which takes current iter (`i` in the code) and returns:

*  For existing `thrust_sort`, always return `n_values`
*  For `thrust_sort_nms`, returns `valid_count_ptr[i]` (`current_values`)

Does this make sense?

File: topi/python/topi/cuda/sort.py [hunk 5]
```diff
@@ -237,6 +237,53 @@ def sort_nms_ir(data, valid_count, output, axis, is_ascend):
...
+    dtype : string, optional
+        DType of the output indices.
+
+    Returns
+    -------
+    out : tvm.te.Tensor
+        The output of this function.
+    """
+    if axis < 0:
+        axis = len(data.shape) + axis
```
Comment (kazum) [inline topi/python/topi/cuda/sort.py]: Add check for the axis.  For example,
```
assert axis==len(data.shape)-1, "Supports sorting along the last axis only"
```

**Comment by masahi** [comment]
@Laurawly Can we have implementation like this? Big if/else blocks should be put inside ` thrust_sort_common`.

```
TVM_REGISTER_GLOBAL("tvm.contrib.thrust.sort")
.set_body([](TVMArgs args, TVMRetValue* ret) {
  CHECK_GE(args.num_args, 5);
  DLTensor* input = args[0];
  DLTensor* valid_count = args[1];
  DLTensor* values_out = args[2];
  DLTensor* indices_out = args[3];
  bool is_ascend = args[4];

  auto data_dtype = DLDataType2String(input->dtype);
  auto out_dtype = DLDataType2String(indices_out->dtype);

  int n_values = input->shape[input->ndim - 1];
  auto get_sort_len = [=](int i) { return n_values; };
  thrust_sort_common(input, values_out, indices_out, is_ascend, get_sort_len, data_type, out_dtype);
}

TVM_REGISTER_GLOBAL("tvm.contrib.thrust.sort_nms")
.set_body([](TVMArgs args, TVMRetValue* ret) {
  CHECK_GE(args.num_args, 5);
  DLTensor* input = args[0];
  DLTensor* valid_count = args[1];
  DLTensor* values_out = args[2];
  DLTensor* indices_out = args[3];
  bool is_ascend = args[4];

  auto data_dtype = DLDataType2String(input->dtype);
  auto out_dtype = DLDataType2String(indices_out->dtype);

  thrust::device_ptr<IndicesType> valid_count_ptr(static_cast<IndicesType *>(valid_count->data));
  auto get_sort_len = [&valid_count_ptr](int i) { return valid_count_ptr[i] };
  thrust_sort_common(input, values_out, indices_out, is_ascend, get_sort_len, data_type, out_dtype);
}

File: src/runtime/contrib/thrust/thrust.cu (hunk 3 above)
Comment (masahi) [reply src/runtime/contrib/thrust/thrust.cu]: I mean we shouldn't duplicate this if/else blocks. See my comment below.

File: topi/python/topi/cuda/sort.py (hunk 5 above)
Comment (Laurawly) [reply topi/python/topi/cuda/sort.py]: We don't only support sorting along the last axis.

Comment (Laurawly) [reply topi/python/topi/cuda/sort.py]: Shall I add `assert 0 <= axis < ndim` after this line?

Comment (Laurawly) [reply topi/python/topi/cuda/sort.py]: In original sort_nms, we do support `axis` other than -1, I'll add your swap operations in this operator.

**Comment by Laurawly** [comment]
> @Laurawly Can we have implementation like this? Big if/else blocks should be put inside ` thrust_sort_common`.
> 
> ```
> TVM_REGISTER_GLOBAL("tvm.contrib.thrust.sort")
> .set_body([](TVMArgs args, TVMRetValue* ret) {
>   CHECK_GE(args.num_args, 5);
>   DLTensor* input = args[0];
>   DLTensor* valid_count = args[1];
>   DLTensor* values_out = args[2];
>   DLTensor* indices_out = args[3];
>   bool is_ascend = args[4];
> 
>   auto data_dtype = DLDataType2String(input->dtype);
>   auto out_dtype = DLDataType2String(indices_out->dtype);
> 
>   int n_values = input->shape[input->ndim - 1];
>   auto get_sort_len = [=](int i) { return n_values; };
>   thrust_sort_common(input, values_out, indices_out, is_ascend, get_sort_len, data_type, out_dtype);
> }
> 
> TVM_REGISTER_GLOBAL("tvm.contrib.thrust.sort_nms")
> .set_body([](TVMArgs args, TVMRetValue* ret) {
>   CHECK_GE(args.num_args, 5);
>   DLTensor* input = args[0];
>   DLTensor* valid_count = args[1];
>   DLTensor* values_out = args[2];
>   DLTensor* indices_out = args[3];
>   bool is_ascend = args[4];
> 
>   auto data_dtype = DLDataType2String(input->dtype);
>   auto out_dtype = DLDataType2String(indices_out->dtype);
> 
>   thrust::device_ptr<IndicesType> valid_count_ptr(static_cast<IndicesType *>(valid_count->data));
>   auto get_sort_len = [&valid_count_ptr](int i) { return valid_count_ptr[i] };
>   thrust_sort_common(input, values_out, indices_out, is_ascend, get_sort_len, data_type, out_dtype);
> }
> ```

Lemme know if the changes look good to you.

File: src/runtime/contrib/thrust/thrust.cu [hunk 6]
```diff
@@ -53,81 +54,112 @@ void thrust_sort(DLTensor* input,
...
+                        DLTensor* values_out,
+                        DLTensor* indices_out,
+                        bool is_ascend,
+                        F &get_sort_len,
+                        std::string data_dtype,
+                        std::string out_dtype) {
   if (data_dtype == "float32") {
     if (out_dtype == "int32") {
-      thrust_sort<float, int32_t>(input, values_out, indices_out, is_ascend);
+      thrust_sort<F, float, int32_t>(input, values_out, indices_out, is_ascend, get_sort_len);
```
Comment (masahi) [inline src/runtime/contrib/thrust/thrust.cu]: I think if you bring `typename F` after `typename, DataType, typename IndicesType` in `thrust_sort`, you can remove explicit `F` in `thrust_sort<F, ...>`

**Comment by masahi** [comment]
> Lemme know if the changes look good to you

Yes this looks great! Thanks

File: src/runtime/contrib/thrust/thrust.cu [hunk 7]
```diff
@@ -53,81 +54,114 @@ void thrust_sort(DLTensor* input,
   thrust::copy(data_ptr, data_ptr + n_iter * n_values, values_ptr);
 
   for (int i = 0 ; i < n_iter; ++i) {
-    thrust::sequence(indices_ptr, indices_ptr + n_values);
+    int current_values = get_sort_len(i);
```
Comment (masahi) [inline src/runtime/contrib/thrust/thrust.cu]: I prefer renaming `current_values` to `n_values` here

File: src/runtime/contrib/thrust/thrust.cu [hunk 8]
```diff
@@ -35,11 +35,12 @@ namespace contrib {
...
 
 // Performs sorting along axis -1 and returns both sorted values and indices.
-template<typename DataType, typename IndicesType>
+template<typename DataType, typename IndicesType, typename F>
 void thrust_sort(DLTensor* input,
                  DLTensor* out_values,
                  DLTensor* out_indices,
-                 bool is_ascend) {
+                 bool is_ascend,
+                 const F &get_sort_len) {
```
Comment (kazum) [inline src/runtime/contrib/thrust/thrust.cu]: I think `std::function<int(int)>` looks easier to read than introducing typename F.

**Comment by masahi** [comment]
Thanks @Laurawly @kazum

**Pull request merged**